File: python/structures/src/searches/graph/dfs.py
def traverse(root):
    output = ''

    stack = [root]
    visited = []

    while len(stack) > 0:
        # pop the next item off the stack
        curr = stack.pop()

        # add all unvisited neighbors to the queue
        for neighbor in curr.neighbors:
            if neighbor not in visited and neighbor not in stack:
                stack.append(neighbor)

        output += str(curr.value) + ' '

        visited.append(curr)

    return output.strip()

File: ruby/src/searches/graph/dfs.rb
require_relative '../../graph/directed'

module DepthFirst
    def DepthFirst.traverse(root)
        out = ''
        stack = [root]
        visited = []

        while !stack.empty?
            curr = stack.pop

            out += curr.value.to_s + ' '

            for neighbor in curr.neighbors
                if !stack.include?(neighbor) && !visited.include?(neighbor)
                    stack << neighbor
                end
            end

            visited << curr
        end

        return out.strip
    end
end
